GHCup integration

bradrn opened the pull request:
Fixes #10. No need to merge just yet; I’ll be pushing more commits as I progressively improve the UI and functionality. (At the moment, it has only the bare minimum required to prove that integration actually works.)

bradrn committed:
Add initial integration with GHCup library

Specifically, the list of available versions is now taken straight
from GHCup's package list. Further features are not yet implemented.

This commit also adds a scrollbar on the feature list, simply because
the GUI is unusable without one!

bradrn committed:
Improve display of tool info

bradrn committed:
Filter out old tool versions

bradrn committed:
Hardcode GHCup logs to stdout

It turns out that GHCup hardcodes its console output to use
ANSI escape sequences (via 'pretty-terminal'). Thus, for now it seems
useless to output these logs to anything other than a console. It
would be nice to instead redirect these to e.g. a graphical toast, but
for now it seems the GUI needs to handle such things itself.

bradrn committed:
Don't forget to auto-format with fourmolu!

Kleidukos commented on src/UI/GHCup.hs:
Does this really belong to the UI layer? I'd says it rather belongs in a Driver/ directory, since it handles interactions with lib:ghcup?

bradrn replied:
Fair enough. I’ve just moved it over.

(I thought I already posted a minute ago, but apparently it deleted my comment when I pressed ‘Resolve’…)

bradrn committed:
Move GHCup initialisation from UI.* to Driver.*

bradrn committed:
Add support for installing/removing GHC

Other tools are not yet supported. This has to be compiled with
-threaded, otherwise lib:ghcup hangs; the issue seems to be that it
forks a thread to write to stdout.

Kleidukos commented on src/Driver/GHCup.hs:
```diff
@@ -133,3 +139,48 @@ initAppState = do
+
+installGHC :: AppState -> GHCupDownloads -> Version -> IO GHCupResult
+installGHC st ds v =
+  ( flip runReaderT st
+      . runResourceT
+      . runE
+      $ installGHCBin v GHCupInternal False []
+  )
+    >>= \case
```
Does the usage of `(&)` help with getting rid of `flip`?

bradrn replied:
I’m not quite sure how that would work… possibly `(runReader & st) . …`, but I’m not sure that’s any more readable. Also, I consider `flip runReaderT` a fairly common idiom for running a monad transformer.

Kleidukos commented on src/UI/Compiler.hs:
```diff
@@ -53,7 +58,8 @@ toActionRow toastOverlay app toolLabel ListResult{..} = do
       ]
   on installButton #stateSet $ \state -> do
-    mockInstall installButton toastOverlay app state (toolLabel <> " " <> versionLabel)
+    let ds = _ghcupDownloads $ ghcupInfo appState
```
How much is that is record access? Can this use record dot access?

Kleidukos commented on src/UI/Install.hs:
```diff
@@ -37,22 +52,38 @@ mockInstall selfSwitch toastOverlay app state toolDescription = do
+            presentToast "Uninstalling…"
+            case tool of
+              GHC -> do
+                rmGHC st ds version >>= \case
```
I like the bind-lambda case pattern :)

Kleidukos commented:
@bradrn can I bother you a bit for this PR? Just remove the freeze file and it should be good for merging
